src/dodal/devices/areadetector/adaravis.py

```diff
@@ -1,99 +1,186 @@
-
-        # Now calling the super method should set the acquire period
-        super(AdAravisDetector, self).stage(*args, **kwargs)
-
-    def _prime_hdf(self) -> None:
-        """
-        Take a single frame and pipe it through the HDF5 writer plugin
-        """
-
-        # Backup state and ensure we are not acquiring
-        reset_to = {signal: signal.get() for signal in self._priming_settings.keys()}
-        self.cam.acquire.set(0).wait(timeout=10)
-
-        # Apply all settings for acquisition
-        for signal, value in self._priming_settings.items():
-            # Ensure that .wait really will wait until the PV is set including its RBV
-            signal.set(value).wait(timeout=10)
-
-        # Acquire a frame
-        self.cam.acquire.set(1).wait(timeout=10)
-
-        # Revert settings to previous values
-        for signal, value in reversed(reset_to.items()):
-            signal.set(value).wait(timeout=10)
+            image_mode = ImageMode.multiple
+        if exposure is not None:
+            await self._drv.acquire_time.set(exposure)
+
+        # trigger mode must be set first and on it's own!
+        await self._drv.trigger_mode.set(ADAravisTriggerMode.on)
+
+        await asyncio.gather(
+            self._drv.trigger_source.set(
+                _get_trigger_source(trigger, self.gpio_number)
+            ),
+            self._drv.num_images.set(num),
+            self._drv.image_mode.set(image_mode),
+        )
+
+        status = await set_and_wait_for_value(self._drv.acquire, True)
+        return status
+
+    async def disarm(self):
+        await stop_busy_record(self._drv.acquire, False, timeout=1)
+
+
+class AdAravisDetector(StandardDetector, HasHints):
+    """
+    Ophyd-async implementation of an AdAravis Detector.
+    The detector may be configured for an external trigger on a GPIO port,
+    which must be done prior to attempting to trigger the detector.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        prefix: str,
+        directory_provider: DirectoryProvider,
+        gpio_number: AdAravisController.GPIO_NUMBER = 1,
+        **scalar_sigs: str,
+    ):
+        drv = AdAravisDriver(prefix + "DET:")
+        hdf = NDFileHDF(prefix + "HDF5:")
+
+        self.drv = drv
+        self.hdf = hdf
+        self.stats = NDPluginStats(prefix + "STAT:")
+
+        super().__init__(
+            AdAravisController(drv, gpio_number=gpio_number),
+            HDFWriter(
+                hdf,
+                directory_provider,
+                lambda: self.name,
+                ADBaseShapeProvider(drv),
+                **scalar_sigs,
+            ),
+            config_sigs=(drv.acquire_time, drv.acquire),
+            name=name,
+        )
+
+    def get_external_trigger_gpio(self):
+        return self.controller.gpio_number
+
+    def set_external_trigger_gpio(self, gpio_number: AdAravisController.GPIO_NUMBER):
+        assert gpio_number in get_args(AdAravisController.GPIO_NUMBER)
```

Comment:
This is a bit fiddly - asserting on introspection on an aliased type-hint... I'm not saying it's wrong but maybe could use a brief comment?